api/handlers.go

```diff
@@ -165,3 +171,404 @@ func (api *RecipeAPI) AddRecipeHandler(w http.ResponseWriter, req *http.Request)
+		return
+	}
+
+	// Set the non-updated fields to values of the currentRecipe - this needs to be done otherwise all fields in array will be overwritten
+	updates.DatasetID = instanceID
+	if updates.Editions == nil || len(updates.Editions) == 0 {
+		updates.Editions = currentRecipe.OutputInstances[instanceIndex].Editions
+	}
+	if updates.Title == "" {
+		updates.Title = currentRecipe.OutputInstances[instanceIndex].Title
+	}
+	if updates.CodeLists == nil || len(updates.CodeLists) == 0 {
+		updates.CodeLists = currentRecipe.OutputInstances[instanceIndex].CodeLists
+	}
+
+	// Update Recipe to Mongo
+	update := bson.M{"$set": bson.M{"output_instances." + strconv.Itoa(instanceIndex): updates}}
+	err = api.dataStore.Backend.UpdateRecipe(recipeID, update)
+	if err != nil {
+		log.Event(ctx, "error updating recipe to mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Marshal to output the newly updated recipe
+	updatedRecipe, _ := api.dataStore.Backend.GetRecipe(recipeID)
+	output, err := json.Marshal(updatedRecipe)
+	if err != nil {
+		log.Event(ctx, "error marshaling updated recipe from mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("content-type", "application/json")
+	w.Write(output)
+}
+
+//UpdateCodelistHandler - update specific recipe by ID
+func (api *RecipeAPI) UpdateCodelistHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	//Get Update Recipe ID
+	vars := mux.Vars(req)
+	recipeID := vars["id"]
+	instanceID := vars["instance_id"]
+	codelistID := vars["codelist_id"]
+	logD := log.Data{"recipe_id": recipeID, "instance_id": instanceID, "codelist_id": codelistID}
+
+	// Read body
+	b, err := ioutil.ReadAll(req.Body)
+	defer req.Body.Close()
+	if err != nil {
+		log.Event(ctx, "error in reading request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Unmarshal to the shape of Response struct
+	var updates recipe.CodeList
+	err = json.Unmarshal(b, &updates)
+	if err != nil {
+		log.Event(ctx, "error returned from json unmarshal", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Validating fields of codelist given in request body
+	err = updates.ValidateUpdateCodeList()
+	if err != nil {
+		log.Event(ctx, "bad request error for invalid updates given in request body", log.ERROR, log.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Find index of specific codelist and interface of codelist in output_instances of the recipe
+	currentRecipe, err := api.dataStore.Backend.GetRecipe(recipeID)
+	if err != nil {
+		log.Event(ctx, "error retrieving specific recipe from mongo", log.ERROR, log.Error(err), logD)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	instanceIndex := -1
+	codelistIndex := -1
+	for i, instance := range currentRecipe.OutputInstances {
```

Comment:
this 'find instance' and 'find codelist' logic, you might consider refactoring into a separate function. Particualrly in this case, its adding a lot to this handler because of the double check we're performing. But the more simple check that exists in other handlers just for the instance is repeated a few times, and might benefit from being shared.

Im imagining a function with a signature roughly `func findInstance(list recipe.OutputInstances) (int, error)` where we either return positive integers or the error, so in this case we'd only have to do 1 `if err != nil > StatusInternalServerError`

Reply:
Done in commit "move bson statements to mongo package - mongo package encapsulation"